### PR TITLE
Add type hints for p2p package and enable stricter checks

### PR DESCRIPTION
**Pull request by cburgdorf**
### What was wrong?

P2P was missing many type hints and didn't have the `--disallow-untyped-defs` check yet which means that new code could potentially land without type hints (making the fight for 100 % coverage a bit of a fight against windmills)

### How was it fixed?

Added type hints and enabled `--disallow-untyped-defs` check

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: p2p/chain.py [hunk 1]
```diff
@@ -403,7 +404,9 @@ def _handle_block_headers(self, headers: List[BlockHeader]) -> None:
         # the executor when the list() is applied to it.
         receipts_tries = await self.wait(loop.run_in_executor(self._executor, list, iterator))
         downloaded: List[DownloadedBlockPart] = []
-        for (receipts, (receipt_root, trie_dict_data)) in zip(receipts_by_block, receipts_tries):
+        # TODO: figure out why mypy is losing the type of the receipts_tries
+        # so we can get rid of the ignore
```
Comment (gsalgado) [inline p2p/chain.py]: I'm not sure what this means. Shouldn't mypy treat receipts_tries's type as Any since that's what self.wait() returns?

File: p2p/discovery.py [hunk 2]
```diff
@@ -125,25 +128,23 @@ def _get_handler(self, cmd: Command) -> Callable[[kademlia.Node, List[Any], byte
...
 
-    def _get_max_neighbours_per_packet(self):
+    def _get_max_neighbours_per_packet(self) -> int:
         if self._max_neighbours_per_packet_cache is not None:
             return self._max_neighbours_per_packet_cache
         self._max_neighbours_per_packet_cache = _get_max_neighbours_per_packet()
         return self._max_neighbours_per_packet_cache
 
-    def connection_made(self, transport):
+    def connection_made(self, transport: Any) -> None:
```
Comment (gsalgado) [inline p2p/discovery.py]: Can't we use an actual transport type here?

File: p2p/eth.py [hunk 3]
```diff
@@ -108,7 +108,7 @@ class ETHProtocol(Protocol):
     cmd_length = 17
     logger = logging.getLogger("p2p.eth.ETHProtocol")
 
-    def send_handshake(self, head_info):
+    def send_handshake(self, head_info: Any) -> None:
```
Comment (gsalgado) [inline p2p/eth.py]: Why use Any here?

File: p2p/les.py [hunk 4]
```diff
@@ -86,7 +95,7 @@ def decode_payload(self, rlp_data: bytes) -> Generator[Tuple[str, Any], None, No
             else:
                 yield key, value
 
-    def encode_payload(self, data):
+    def encode_payload(self, data: Dict[str, bytes]) -> bytes:  # type: ignore
```
Comment (gsalgado) [inline p2p/les.py]: Why do we need the `type: ignore` comment here?

File: p2p/les.py [hunk 5]
```diff
@@ -254,7 +263,7 @@ class LESProtocol(Protocol):
     _commands = [Status, Announce, BlockHeaders, BlockBodies, Receipts, Proofs, ContractCodes]
     cmd_length = 15
 
-    def send_handshake(self, head_info):
+    def send_handshake(self, head_info: Any) -> None:
```
Comment (gsalgado) [inline p2p/les.py]: Should be able to use `HeadInfo` here, no?

File: p2p/les.py [hunk 6]
```diff
@@ -354,7 +363,7 @@ class LESProtocolV2(LESProtocol):
     _commands = [StatusV2, Announce, BlockHeaders, BlockBodies, Receipts, ProofsV2, ContractCodes]
     cmd_length = 21
 
-    def send_handshake(self, head_info):
+    def send_handshake(self, head_info: Any) -> None:
```
Comment (gsalgado) [inline p2p/les.py]: And here

File: p2p/lightchain.py [hunk 7]
```diff
@@ -257,14 +260,14 @@ def register_peer(self, peer: BasePeer) -> None:
             # TODO push validation into process pool executor
             VM.validate_header(header, parent_header)
 
-    async def _cleanup(self):
+    async def _cleanup(self) -> None:
         self.logger.info("Stopping LightPeerChain...")
 
-    async def _wait_for_reply(self, request_id: int) -> Dict[str, Any]:
+    async def _wait_for_reply(self, request_id: int) -> Any:
```
Comment (gsalgado) [inline p2p/lightchain.py]: Why this change?

File: p2p/peer.py [hunk 8]
```diff
@@ -238,11 +239,12 @@ def remove_subscriber(self, subscriber: 'asyncio.Queue[PEER_MSG_TYPE]') -> None:
         if subscriber in self._subscribers:
             self._subscribers.remove(subscriber)
 
-    async def do_sub_proto_handshake(self):
+    async def do_sub_proto_handshake(self) -> None:
         """Perform the handshake for the sub-protocol agreed with the remote peer.
 
         Raises HandshakeFailure if the handshake is not successful.
         """
+        self.base_protocol = cast(P2PProtocol, self.base_protocol)
```
Comment (gsalgado) [inline p2p/peer.py]: `BasePeer.__init__()` does `self.base_protocol = P2PProtocol(self)`, so I'm wondering why this is needed

File: p2p/peer.py [hunk 9]
```diff
@@ -258,11 +260,12 @@ def remove_subscriber(self, subscriber: 'asyncio.Queue[PEER_MSG_TYPE]') -> None:
         await self.process_sub_proto_handshake(cmd, msg)
         self.logger.debug("Finished %s handshake with %s", self.sub_proto, self.remote)
 
-    async def do_p2p_handshake(self):
+    async def do_p2p_handshake(self) -> None:
         """Perform the handshake for the P2P base protocol.
 
         Raises HandshakeFailure if the handshake is not successful.
         """
+        self.base_protocol = cast(P2PProtocol, self.base_protocol)
```
Comment (gsalgado) [inline p2p/peer.py]: Ditto

File: p2p/peer.py [hunk 10]
```diff
@@ -347,7 +350,7 @@ def close(self):
                 self.logger.debug("%s disconnected: %s", self, e)
                 return
 
-    async def read_msg(self) -> Tuple[protocol.Command, protocol._DecodedMsgType]:
+    async def read_msg(self) -> Tuple[protocol.Command, Dict[str, Any]]:
```
Comment (gsalgado) [inline p2p/peer.py]: Why this change?  This method can return things other than a dict, that's why it used `protocol._DecodedMsgType`

File: p2p/peer.py [hunk 11]
```diff
@@ -361,12 +364,13 @@ def close(self):
...
         # ProcessPoolExecutor(). Need to make sure we don't use all CPUs in the machine for that,
         # though, otherwise asyncio's event loop can't run and we can't keep up with other peers.
-        decoded_msg = cmd.decode(msg)
+        decoded_msg = cast(Dict[str, Any], cmd.decode(msg))
         self.logger.debug("Successfully decoded %s msg: %s", cmd, decoded_msg)
         return cmd, decoded_msg
 
     def handle_p2p_msg(self, cmd: protocol.Command, msg: protocol._DecodedMsgType) -> None:
         """Handle the base protocol (P2P) messages."""
+        self.base_protocol = cast(P2PProtocol, self.base_protocol)
```
Comment (gsalgado) [inline p2p/peer.py]: Another cast I wouldn't think to be needed.

File: p2p/peer.py [hunk 12]
```diff
@@ -478,6 +482,7 @@ def disconnect(self, reason: DisconnectReason) -> None:
 
         :param reason: An item from the DisconnectReason enum.
         """
+        self.base_protocol = cast(P2PProtocol, self.base_protocol)
```
Comment (gsalgado) [inline p2p/peer.py]: And here too

File: p2p/discovery.py (hunk 2 above)
Comment (cburgdorf) [reply p2p/discovery.py]: Ah, good hint. It's one of these things that I skipped and never came back to. So this represents something like a `asyncio.DatagramTransport` right?

File: p2p/les.py (hunk 5 above)
Comment (cburgdorf) [reply p2p/les.py]: Ah, good point. I skipped that because I didn't notice `HeadInfo` actually exists.

Comment (cburgdorf) [reply p2p/les.py]: Ah, turns out it's actually `ChainInfo` there (because `genesis_hash` doesn't exist on `HeadInfo`) 

https://github.com/ethereum/py-evm/pull/947/commits/69892c3ff720ad437d76ca29d1c94df81c5380e1

Notice, I'll rebase these changes into the right place once this passed a complete review.

File: p2p/les.py (hunk 6 above)
Comment (cburgdorf) [reply p2p/les.py]: Both places fixed.

File: p2p/chain.py (hunk 1 above)
Comment (cburgdorf) [reply p2p/chain.py]: [`wait(...)`](https://github.com/cburgdorf/py-evm/blob/69892c3ff720ad437d76ca29d1c94df81c5380e1/p2p/service.py#L45-L52) is generic now and should return the right type which works in principle but fails at this place.

File: p2p/discovery.py (hunk 2 above)
Comment (cburgdorf) [reply p2p/discovery.py]: Fixed by https://github.com/ethereum/py-evm/pull/947/commits/87d585801d1da8b99dbe60595299dab5734c671f

File: p2p/peer.py (hunk 10 above)
Comment (cburgdorf) [reply p2p/peer.py]: True, some temporary change that made it through my own review.

File: p2p/les.py (hunk 4 above)
Comment (cburgdorf) [reply p2p/les.py]: It doesn't conform to the signature of the super type.

We could also do this:
```
    def encode_payload(self, data: Union[_DecodedMsgType, sedes.CountableList]) -> bytes:
        response = [
            (key, self.items_sedes[key].serialize(value))
            for key, value
            in sorted(data.items())  # type: ignore
        ]
        return super().encode_payload(response)
```
So, we conform to the interface but we then ignore that `data.items()` is actually invalid because `items()` doesn't exist on a Tuple.

File: p2p/lightchain.py (hunk 7 above)
Comment (cburgdorf) [reply p2p/lightchain.py]: Good point, fixed and explained here https://github.com/ethereum/py-evm/pull/947/commits/6cf408c432c74f78d2029ced398774a6c85f142d

File: p2p/peer.py (hunk 8 above)
Comment (cburgdorf) [reply p2p/peer.py]: Yep, we can do better here. Checkout https://github.com/ethereum/py-evm/pull/947/commits/a8632048824259f6816b57aca92c7c72d8c36d24

Basically we:

- get rid of the type of `self.base_protocol` on the class level so that it is really only defined by `__init__`
- return early in `get_protocol_command_for` so that the intermediate variable doesn't have to be types as `Union[Protocol, P2PProtocol]`

File: p2p/peer.py (hunk 9 above)
Comment (cburgdorf) [reply p2p/peer.py]: solved, see previous comment

File: p2p/peer.py (hunk 11 above)
Comment (cburgdorf) [reply p2p/peer.py]: solved, see previous comment

File: p2p/peer.py (hunk 12 above)
Comment (cburgdorf) [reply p2p/peer.py]: solved, see previous comment

File: p2p/eth.py (hunk 3 above)
Comment (cburgdorf) [reply p2p/eth.py]: Fixed by https://github.com/ethereum/py-evm/pull/947/commits/7ba3b72ffe102aec18bf3095159841326ac79afe

**Comment by cburgdorf** [comment]
@gsalgado Thanks for the sharp review. I was able to tighten types for a lot of places. I think I addressed all your comments. Once you approve, I'll make sure to move all the wip commits into the first one before merging.

That said, CircleCI seems to have hiccups right now so this won't go green until Circle runs smoothly again.

File: p2p/peer.py (hunk 10 above)
Comment (gsalgado) [reply p2p/peer.py]: Are you going to revert it?

Comment (cburgdorf) [reply p2p/peer.py]: Sorry, fixed by https://github.com/ethereum/py-evm/pull/947/commits/1bbab42b47e15f6453894dcc993e1d9bab925a05

**Comment by cburgdorf** [comment]
Oh, just noticed its not CircleCI that has a hiccup, it's actually blocked by https://github.com/ethereum/py-evm/issues/953
